### ISSUE TITLE
Infinite types can cause stack overflow

### ISSUE THREAD

**Issue by jcollard** [elm-lang/elm-compiler#832]
**Other Examples:** #707, #956, #983, #820, #987

The following code produces an infinite type that isn't detected.

```haskell
foo b =
    if | True -> b b
       | 0 >= b -> False
```
This originated from this issue: https://github.com/elm-lang/elm-repl/issues/45

This is the smallest I could get the example while still producing a stack overflow.

Changing the order of the branches causes a type error (as expected).

It looks like the `0 >= b` causes `b :: comparable` this doesn't rule out it being a function type. When we then hit `b b`, it seems to infer `comparable -> (comparable -> (comparable -> ....` which results in a stack overflow.

This explanation it purely based of intuition as I'm not at all familiar with what is actually going on with the type inference.

**Comment by Dedoig** [comment]
Just a "me too" on this ... I have hit a variation of this a couple of times where the compiler goes into some infinite loop (but I don't get a stack overflow). I reported the 1st one on the list and was told it was from an incorrect second import statement:

```
module Reverse where

import Html
import StartApp

main : Signal Html.Html
main =
  StartApp.start { model = 0, view = view, update = update }

view addr m = Html.text "Hello"

update x m = m
```

I did not record the second one, but it was not related to a faulty import. It did thoroughly lock up my entire Mac and I had to claw through to a terminal for some "kill -9 ..." treatment.
